Add callback function before/after the initial Esolve

RemiLehe opened the pull request:
This adds new callback functions:
- `beforeInitEsolve`
- `afterInitEsolve`

that are called before/after the Poisson solve that we do before the beginning of the PIC loop.

dpgrote reviewed (approved):
Looks good. Thanks!

RemiLehe commented on Source/Initialization/WarpXInitData.cpp:
```diff
@@ -413,7 +414,9 @@ WarpX::InitData ()
     {
         // Loop through species and calculate their space-charge field
         bool const reset_fields = false; // Do not erase previous user-specified values on the grid
+        ExecutePythonCallback("beforeEsolve");
```
```suggestion
        ExecutePythonCallback("beforeInitEsolve");
```

RemiLehe commented on Source/Initialization/WarpXInitData.cpp:
```diff
@@ -413,7 +414,9 @@ WarpX::InitData ()
         bool const reset_fields = false; // Do not erase previous user-specified values on the grid
+        ExecutePythonCallback("beforeEsolve");
         ComputeSpaceChargeField(reset_fields);
+        ExecutePythonCallback("afterEsolve");
```
```suggestion
        ExecutePythonCallback("afterInitEsolve");
```

dpgrote commented on Python/pywarpx/callbacks.py:
```diff
@@ -352,13 +385,13 @@ def callfrombeforeEsolve(f):
     return f
 def installbeforeEsolve(f):
-    "Adds a function to the list of functions called before an E solve"
+    "Adds a function to the list of functions called before the initial E solve"
```
Were these changes inadvertent?

RemiLehe commented:
Thanks for the quick feedback!
I corrected the accidental changes in the docstring of `afterEsolve`, and updated `WarpX_py.H`. Do you think that this is ready for merging?

dpgrote reviewed (approved):
Looks good. Thanks!

roelof-groenewald reviewed (approved):
Thanks!
